File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/legacy/core/FocusMeteringAction.kt
package dev.hebei.camerax_android.legacy.core

import dev.hebei.camerax_android.legacy.*

class FocusMeteringAction(registrar: CameraXRegistrar) : PigeonApiFocusMeteringAction(registrar) {
    override fun build(
        first: MeteringPointArgs.Stub,
        others: List<MeteringPointArgs.Stub>?,
        disableAutoCancel: Boolean?,
        autoCancelDuration: DurationArgs.Stub?
    ): androidx.camera.core.FocusMeteringAction {
        val builder = if (first.mode == null) androidx.camera.core.FocusMeteringAction.Builder(first.point)
        else androidx.camera.core.FocusMeteringAction.Builder(first.point, first.mode)
        if (others != null) {
            for (otherPoint in others) {
                if (otherPoint.mode == null) builder.addPoint(otherPoint.point)
                else builder.addPoint(otherPoint.point, otherPoint.mode)
            }
        }
        if (disableAutoCancel != null) {
            builder.disableAutoCancel()
        }
        if (autoCancelDuration != null) {
            builder.setAutoCancelDuration(autoCancelDuration.duration, autoCancelDuration.timeUnit)
        }
        return builder.build()
    }

    override fun getAutoCancelDurationInMillis(pigeon_instance: androidx.camera.core.FocusMeteringAction): Long {
        return pigeon_instance.autoCancelDurationInMillis
    }

    override fun getMeteringPointsAe(pigeon_instance: androidx.camera.core.FocusMeteringAction): List<androidx.camera.core.MeteringPoint> {
        return pigeon_instance.meteringPointsAe
    }

    override fun getMeteringPointsAf(pigeon_instance: androidx.camera.core.FocusMeteringAction): List<androidx.camera.core.MeteringPoint> {
        return pigeon_instance.meteringPointsAf
    }

    override fun getMeteringPointsAwb(pigeon_instance: androidx.camera.core.FocusMeteringAction): List<androidx.camera.core.MeteringPoint> {
        return pigeon_instance.meteringPointsAwb
    }

    override fun isAutoCancelEnabled(pigeon_instance: androidx.camera.core.FocusMeteringAction): Boolean {
        return pigeon_instance.isAutoCancelEnabled
    }

    class MeteringPointArgs(registrar: CameraXRegistrar) : PigeonApiMeteringPointArgs(registrar) {
        override fun pigeon_defaultConstructor(
            point: androidx.camera.core.MeteringPoint, modes: List<MeteringMode>?
        ): Stub {
            val mode = if (modes.isNullOrEmpty()) null
            else modes.fold(0) { total, next -> total or next.obj }
            return Stub(point, mode)
        }

        data class Stub(
            val point: androidx.camera.core.MeteringPoint,
            @androidx.camera.core.FocusMeteringAction.MeteringMode val mode: Int?
        )
    }

    class DurationArgs(registrar: CameraXRegistrar) : PigeonApiDurationArgs(registrar) {
        override fun pigeon_defaultConstructor(duration: Long, timeUnit: TimeUnit): Stub {
            return Stub(duration, timeUnit.obj)
        }

        data class Stub(val duration: Long, val timeUnit: java.util.concurrent.TimeUnit)
    }
}

val MeteringMode.obj
    get() = when (this) {
        MeteringMode.AF -> androidx.camera.core.FocusMeteringAction.FLAG_AF
        MeteringMode.AE -> androidx.camera.core.FocusMeteringAction.FLAG_AE
        MeteringMode.AWB -> androidx.camera.core.FocusMeteringAction.FLAG_AWB
    }

val TimeUnit.obj
    get() = when (this) {
        TimeUnit.NANOSECONDS -> java.util.concurrent.TimeUnit.NANOSECONDS
        TimeUnit.MICROSECONDS -> java.util.concurrent.TimeUnit.MICROSECONDS
        TimeUnit.MILLISECONDS -> java.util.concurrent.TimeUnit.MILLISECONDS
        TimeUnit.SECONDS -> java.util.concurrent.TimeUnit.SECONDS
        TimeUnit.MINUTES -> java.util.concurrent.TimeUnit.MINUTES
        TimeUnit.HOURS -> java.util.concurrent.TimeUnit.HOURS
        TimeUnit.DAYS -> java.util.concurrent.TimeUnit.DAYS
    }

File: camerax_android/lib/src/legacy/core/focus_metering_action.dart
import 'package:camerax_android/src/legacy/camerax.g.dart' as $native;
import 'package:camerax_platform_interface/camerax_platform_interface.dart'
    as $interface;

import 'metering_mode.dart';
import 'metering_point.dart';

final class FocusMeteringAction extends $interface.FocusMeteringAction {
  final $native.FocusMeteringAction obj;

  FocusMeteringAction.$native(this.obj) : super.impl();

  factory FocusMeteringAction(
    ($interface.MeteringPoint, List<$interface.MeteringMode>) first, {
    List<($interface.MeteringPoint, List<$interface.MeteringMode>)>? others,
    bool? disableAutoCancel,
    Duration? autoCancelDuration,
  }) {
    final obj = $native.FocusMeteringAction.build(
      first: first.obj,
      others: others?.map((other) => other.obj).toList(),
      disableAutoCancel: disableAutoCancel,
      autoCancelDuration: autoCancelDuration?.obj,
    );
    return FocusMeteringAction.$native(obj);
  }

  @override
  Future<Duration> getAutoCancelDuration() async {
    final inMillis = await obj.getAutoCancelDurationInMillis();
    return Duration(
      milliseconds: inMillis,
    );
  }

  @override
  Future<List<$interface.MeteringPoint>> getMeteringPointsAe() async {
    final objs = await obj.getMeteringPointsAe();
    return objs.map((obj) => obj.args).toList();
  }

  @override
  Future<List<$interface.MeteringPoint>> getMeteringPointsAf() async {
    final objs = await obj.getMeteringPointsAf();
    return objs.map((obj) => obj.args).toList();
  }

  @override
  Future<List<$interface.MeteringPoint>> getMeteringPointsAwb() async {
    final objs = await obj.getMeteringPointsAwb();
    return objs.map((obj) => obj.args).toList();
  }

  @override
  Future<bool> isAutoCancelEnabled() async {
    final value = await obj.isAutoCancelEnabled();
    return value;
  }
}

extension on ($interface.MeteringPoint, List<$interface.MeteringMode>) {
  $native.MeteringPointArgs get obj {
    final point = this.$1;
    if (point is! MeteringPoint) {
      throw TypeError();
    }
    return $native.MeteringPointArgs(
      point: point.obj,
      modes: $2.map((mode) => mode.obj).toList(),
    );
  }
}

extension on Duration {
  $native.DurationArgs get obj {
    return $native.DurationArgs(
      duration: inMilliseconds,
      timeUnit: $native.TimeUnit.milliseconds,
    );
  }
}
